### PR TITLE
gh-128035: Add ssl.HAS_PHA to detect libssl PHA support

### PR DESCRIPTION
**Pull request by WillChilds-Klein**
# Notes

Please see #128035's description.

# Testing
- CPython CI
- confirmed that PHA-related tests no longer fail when CPython is built against AWS-LC

* Issue: gh-128035

----
📚 Documentation preview 📚: https://cpython-previews--128036.org.readthedocs.build/

### CODE REVIEW

File: Doc/library/ssl.rst [hunk 1]
```diff
@@ -934,6 +934,12 @@ Constants
 
    .. versionadded:: 3.13
 
+.. data:: HAS_PHA
+
+   Whether the OpenSSL library has built-in support for TLS post-handshake auth (PHA).
+
+   .. versionadded:: 3.14
```
Comment (tomasr8) [inline Doc/library/ssl.rst]:
```suggestion
   .. versionadded:: next
```

**Review by tomasr8 — COMMENTED** [review]
This will also need a news entry :)

File: Doc/library/ssl.rst [hunk 2]
```diff
@@ -934,6 +934,12 @@ Constants
 
    .. versionadded:: 3.13
 
+.. data:: HAS_PHA
+
+   Whether the OpenSSL library has built-in support for TLS post-handshake auth (PHA).
```
Comment (picnixz) [inline Doc/library/ssl.rst]:
```suggestion
   Whether the OpenSSL library has built-in support for TLS-PHA.
```

We don't specify what TLS-PSK is, so I don't think we need to specify what PHA is.

File: Lib/test/test_httplib.py [hunk 3]
```diff
@@ -2073,8 +2073,8 @@ def test_host_port(self):
 
     def test_tls13_pha(self):
         import ssl
-        if not ssl.HAS_TLSv1_3:
-            self.skipTest('TLS 1.3 support required')
+        if not ssl.HAS_TLSv1_3 or not ssl.HAS_PHA:
+            self.skipTest('TLS 1.3 post-handshake auth (PHA) support required')
```
Comment (picnixz) [inline Lib/test/test_httplib.py]:
```suggestion
            self.skipTest('TLS 1.3 PHA support required')
```

File: Misc/NEWS.d/next/Core_and_Builtins/2024-12-17-18-20-37.gh-issue-128035.JwqHdB.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+TLSv1.3 post-handshake client authentication (PHA), often referred to as "mutual TLS" or "mTLS", allows TLS servers to authenticate client identities using digital certificates. This commit exposes a boolean property ``ssl.HAS_PHA`` to indicate whether the crypto library CPython is built against supports PHA, allowing python's test suite and consuming modules to branch accordingly.
```
Comment (picnixz) [inline Misc/NEWS.d/next/Core_and_Builtins/2024-12-17-18-20-37.gh-issue-128035.JwqHdB.rst]: A shorter NEWS should be written. The NEWS is a message that users will see (it's in the changelog). Some suggestion:

```rst
Indicate through :data:`ssl.HAS_PHA` whether the :mod:`ssl` module supports TLSv1.3
post-handshake client authentication (PHA). Patch by YOURNAME.
```

In addition, you should add a What's New entry in `Doc/whatsnew/3.14.rst` indicating the additional constant. Usually, the same message as for the NEWS entry can be reused (check the other entries for the formatting).

Comment (WillChilds-Klein) [reply Misc/NEWS.d/next/Core_and_Builtins/2024-12-17-18-20-37.gh-issue-128035.JwqHdB.rst]: Thank you for the guidance. I've updated the news and whatsnew files accordingly.

File: Doc/whatsnew/3.14.rst [hunk 5]
```diff
@@ -70,6 +70,9 @@ Summary -- Release highlights
 New Features
 ============
 
+* Indicate through :data:`ssl.HAS_PHA` whether the :mod:`ssl` module supports TLSv1.3
+  post-handshake client authentication (PHA). (Contributed by Will Childs-Klein in
+  :gh:`128036`.)
```
Comment (tomasr8) [inline Doc/whatsnew/3.14.rst]: This should go under `Improved modules`. You can make a new section for `ssl`:

```rst
ssl
---

* Indicate through :data:`ssl.HAS_PHA` whether the :mod:`ssl` module supports TLSv1.3
  post-handshake client authentication (PHA). (Contributed by Will Childs-Klein in
  :gh:`128036`.)
```

File: Doc/whatsnew/3.14.rst [hunk 6]
```diff
@@ -580,6 +580,14 @@ pydoc
   (Contributed by Jelle Zijlstra in :gh:`101552`.)
 
 
+ssl
+---
+
+* Indicate through :data:`ssl.HAS_PHA` whether the :mod:`ssl` module supports
+  TLSv1.3 post-handshake client authentication (PHA).
```
Comment (picnixz) [inline Doc/whatsnew/3.14.rst:587]:
```suggestion
* Indicate through :data:`ssl.HAS_PHA` whether the :mod:`ssl` module
  supports TLSv1.3 post-handshake client authentication (PHA).
```

(just for reducing the line length and make it look more uniform, but this is entirely optional and a matter of taste; my previous suggestion may even have been incorrect in that regard).

**Review by picnixz — COMMENTED** [review]
Considering the precedent on HAS_PSK and conditional blake2 detection for OpenSSL-like implementations that do not support it, this an ok change. 

Some questions:

- is there an equivalent of `OPENSSL_NO_PHA` constant or not? (AFAIK, no, but are there some implementations that offer TLSv1.3 without PHA and would define OPENSSL_NO_PHA?)
- is there a better constant than `SSL_VERIFY_POST_HANDSHAKE` to check if PHA is supported or not? namely, is there an implementation which supports PHA (whether by default or not) but uses another constant?

cc @gpshead for a second look as well

**Review by gpshead — APPROVED** [review]
This makes sense to me. If there are better ways to determine this in the future, we can adapt to those. As with a lot of things `libssl` related... I feel like it can be configured via the environment in such ways that regardless of compile time checks capabilities can be changed at runtime via its configs. This constant only covers if the library _could_ support it, but isn't able to express if it happens to be enabled or not within the current process. That is fine and is the expected simple thing to do.

**Pull request merged**